### PR TITLE
Refactor work units and added tests for command.go

### PR DESCRIPTION
**Pull request by resoluteCoder**

### CODE REVIEW

File: pkg/workceptor/command.go [hunk 1]
```diff
@@ -11,24 +11,53 @@ import (
...
 	"syscall"
 	"time"
 
 	"github.com/ghjm/cmdline"
 	"github.com/google/shlex"
 )
 
+type BaseWorkUnitForWorkUnit interface {
+	CancelContext()
+	Debug(format string, v ...interface{})
```
Comment (AaronH88) [inline pkg/workceptor/command.go]: The logger appears to be coupled with the Base work unit here. I wonder if their is a good way to separate them

File: pkg/workceptor/command_test.go [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+	for _, testCase := range paramsTestCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			testCase.expectedCalls()
+			err := wu.SetFromParams(testCase.params)
+			testCase.errorCatch(err, t)
+		})
+	}
+}
+
+func TestUnredactedStatus(t *testing.T) {
```
Comment (AaronH88) [inline pkg/workceptor/command_test.go:91]: Can I get a few more tests in the block please and run them in `t.Parallel` . I would like to make sure that the new status lock wrappers you created do in fact lock the status object for running in parallel

File: pkg/workceptor/command_test.go [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+func statusExpectCalls(mockBaseWorkUnit *mock_workceptor.MockBaseWorkUnitForWorkUnit) {
+	statusLock := &sync.RWMutex{}
+	mockBaseWorkUnit.EXPECT().GetStatusLock().Return(statusLock).Times(2)
+	mockBaseWorkUnit.EXPECT().GetStatusWithoutExtraData().Return(&workceptor.StatusFileData{})
+	mockBaseWorkUnit.EXPECT().GetStatusCopy().Return(workceptor.StatusFileData{
+		ExtraData: &workceptor.CommandExtraData{},
+	})
+}
+
+func createTestSetup(t *testing.T) (workceptor.WorkUnit, *mock_workceptor.MockBaseWorkUnitForWorkUnit, *mock_workceptor.MockNetceptorForWorkceptor, *gomock.Controller, *workceptor.Workceptor) {
```
Comment (AaronH88) [inline pkg/workceptor/command_test.go]: the `ctrl` object returned is never used in your tests. You can drop it

File: pkg/workceptor/command_test.go [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+
+	cwc := &workceptor.CommandWorkerCfg{}
+	mockBaseWorkUnit.EXPECT().Init(w, "", "", workceptor.FileSystem{}, nil)
+	workUnit := cwc.NewWorker(mockBaseWorkUnit, w, "", "")
+
+	return workUnit, mockBaseWorkUnit, mockNetceptor, ctrl, w
+}
+
+func TestCommandSetFromParams(t *testing.T) {
+	wu, mockBaseWorkUnit, _, _, _ := createTestSetup(t) //nolint:dogsled
```
Comment (AaronH88) [inline pkg/workceptor/command_test.go]: interesting lint error, any idea why this has appeared? Will pinning the linter to a version help?

File: pkg/workceptor/kubernetes.go [hunk 5]
```diff
@@ -39,6 +39,7 @@ import (
 // kubeUnit implements the WorkUnit interface.
 type kubeUnit struct {
 	BaseWorkUnit
+	// BaseWorkUnitForWorkUnit
```
Comment (AaronH88) [inline pkg/workceptor/kubernetes.go]: So this is just implemented for the command work unit and kube is next?

File: pkg/workceptor/remote_work.go [hunk 6]
```diff
@@ -23,13 +23,14 @@ import (
 
 // remoteUnit implements the WorkUnit interface for the Receptor remote worker plugin.
 type remoteUnit struct {
-	BaseWorkUnit
+	// BaseWorkUnit
```
Comment (AaronH88) [inline pkg/workceptor/remote_work.go]: remove please

File: pkg/workceptor/remote_work.go [hunk 7]
```diff
@@ -23,13 +23,14 @@ import (
...
+	BaseWorkUnitForWorkUnit
 	topJC  *utils.JobContext
 	logger *logger.ReceptorLogger
 }
 
-// remoteExtraData is the content of the ExtraData JSON field for a remote work unit.
-type remoteExtraData struct {
+// RemoteExtraData is the content of the ExtraData JSON field for a remote work unit.
+type RemoteExtraData struct {
```
Comment (AaronH88) [inline pkg/workceptor/remote_work.go:32]: I can see a lot of changes in remote_work.go here which is grand, but the gap in tests for some of these changes make me nervous.

A lot of the changes are fine, your accessing the logger in a new way. but some of the changes involve changing how we access the `status lock` and casting using the exported `RemoteExtraData` object. 
I would feel a lot more comfortable if these critical paths were tested in this PR.

File: pkg/workceptor/remote_work.go [hunk 8]
```diff
@@ -85,14 +86,14 @@ func (rw *remoteUnit) getConnection(ctx context.Context) (net.Conn, *bufio.Reade
 		if err == nil {
 			return conn, reader
 		}
-		rw.w.nc.GetLogger().Debug("Connection to %s failed with error: %s",
-			rw.Status().ExtraData.(*remoteExtraData).RemoteNode, err)
+		rw.GetWorkceptor().nc.GetLogger().Debug("Connection to %s failed with error: %s",
+			rw.Status().ExtraData.(*RemoteExtraData).RemoteNode, err)
```
Comment (AaronH88) [inline pkg/workceptor/remote_work.go:89]: Test lines like this please

File: pkg/workceptor/remote_work.go [hunk 9]
```diff
@@ -185,7 +186,7 @@ func (rw *remoteUnit) startRemoteUnit(ctx context.Context, conn net.Conn, reader
 	}
 	red.RemoteUnitID = string(match[1])
 	rw.UpdateFullStatus(func(status *StatusFileData) {
-		ed := status.ExtraData.(*remoteExtraData)
+		ed := status.ExtraData.(*RemoteExtraData)
```
Comment (AaronH88) [inline pkg/workceptor/remote_work.go:188]: Test lines like this please

File: pkg/workceptor/remote_work.go [hunk 10]
```diff
@@ -528,10 +529,10 @@ func (rw *remoteUnit) Status() *StatusFileData {
 
 // UnredactedStatus returns a copy of the status currently loaded in memory, including secrets.
 func (rw *remoteUnit) UnredactedStatus() *StatusFileData {
-	rw.statusLock.RLock()
-	defer rw.statusLock.RUnlock()
-	status := rw.getStatus()
-	ed, ok := rw.status.ExtraData.(*remoteExtraData)
+	rw.GetStatusLock().RLock()
```
Comment (AaronH88) [inline pkg/workceptor/remote_work.go:531]: Test lines like this please

File: pkg/workceptor/kubernetes.go (hunk 5 above)
Comment (resoluteCoder) [reply pkg/workceptor/kubernetes.go]: Yes kube is next and remote work. This interface will work for all the work types. I can swap BaseWorkUnit for BaseWorkUnitForWorkUnit in this PR

Comment (resoluteCoder) [reply pkg/workceptor/kubernetes.go]: Well it looks like I already did remote_work. I can do kube as well. Wdyt?

File: pkg/workceptor/command.go (hunk 1 above)
Comment (resoluteCoder) [reply pkg/workceptor/command.go]: Are you referring to the Debug, Error, Info, and Warning methods?

File: pkg/workceptor/command_test.go (hunk 4 above)
Comment (resoluteCoder) [reply pkg/workceptor/command_test.go]: Yeah the lint error was due to have 3 blank vars returning from that func. Now that I have eliminated `ctrl` from the return. I will see if that still persists

Comment (resoluteCoder) [reply pkg/workceptor/command_test.go]: Yeah it looks like it worked with just having two blank vars.

File: pkg/workceptor/command.go (hunk 1 above)
Comment (AaronH88) [reply pkg/workceptor/command.go]: Exactly :smile:

File: pkg/workceptor/kubernetes.go (hunk 5 above)
Comment (AaronH88) [reply pkg/workceptor/kubernetes.go]: Im easy, both are good options. Perhaps Kube after will help with your sanity :grin:

File: pkg/workceptor/kubernetes.go [hunk 11]
```diff
@@ -38,7 +38,8 @@ import (
 
 // kubeUnit implements the WorkUnit interface.
 type kubeUnit struct {
-	BaseWorkUnit
+	BaseWorkUnitForWorkUnit
+	// BaseWorkUnit
```
Comment (AaronH88) [inline pkg/workceptor/kubernetes.go]: remove comment please

**Review by AaronH88 — APPROVED** [review]
lgtm

**Pull request merged**